string_indexing/suffix_array.py

```diff
@@ -63,6 +66,223 @@ def induced_sorting(text, n):
   '''Computes suffix array using Nong-Zhang-Chan algorithm'''
   raise NotImplementedError
 
+def _map_string(s):
+  sorted_char_set = sorted(set( c for c in s[1:] ))
+  char_to_ord = { c : i for i, c in enumerate(sorted_char_set, 1) }
+  return [0] + [ char_to_ord[c] for c in s[1:] ] + [0]
+
+def small_large(text, n):
+  Category = enum.Enum("Category", "L S")
+
+  def _small_large(s):
```

Comment:
This is more general, although debatable issue: I was passing n to all functions although I could compute it from text. I was wondering whether using this as a second parameter in _small_large (instead of len(s)) would help readability. But I'd be also glad to be argued otherwise.

Reply:
For me there are two main problems with passing length as a parameter - additional assumptions and redundancy. Those are more of a problem of interface usability and not readability which I think is a higher concern in this case. But I also don't see a drawback for readability in calling "n = len(s)-1" in a first line of functions that use it. Nevertheless I've modified the interface to conform to existing standard.

Reply:
As for the redundancy - I'd agree completely if it were a production code. But as far as this only serves teaching purposes, I think some looseness on this issue may be allowed, and this was basically the reason why I introduced it in the first place: all codes would have a similar one-liner.